lib/sidekiq/scheduled.rb

```diff
@@ -9,29 +9,45 @@ module Scheduled
     SETS = %w[retry schedule]
 
     class Enq
+      LUA_ZPOPBYSCORE = <<~EOS
+        local key, now = KEYS[1], ARGV[1]
+        local jobs = redis.call("zrangebyscore", key, "-inf", now, "limit", 0, 1)
+        if jobs[1] then
+          redis.call("zrem", key, jobs[1])
+          return jobs[1]
+        end
+      EOS
+
+      LUA_ZPOPBYSCORE_SHA = Digest::SHA1.hexdigest(LUA_ZPOPBYSCORE)
```

Comment:
The expected hash is returned by `eval`. We should not be calculating the SHA directly but using the eval result instead.

Reply:
I think `eval` just returns the return value of the script and `SCRIPT LOAD` is the one that returns a SHA.

We could do that too, though the current code is just being optimistic that the code is already loaded because in most cases it will be. And even if we did `SCRIPT LOAD` at the start here and memoize it, there's no assurance that the script lives in Redis throughout this process' lifetime, so we'd need the error handling / rescue anyway.

I think this is a common pattern also seen in other open-source projects like Discourse: https://github.com/discourse/discourse/blob/1472e47aae5bfdfb6fd9abfe89beb186c751f514/lib/rate_limiter.rb#L198-L206

The Redis docs also says:

> The client library implementation can always optimistically send EVALSHA under the hood even when the client actually calls EVAL, in the hope the script was already seen by the server. If the NOSCRIPT error is returned EVAL will be used instead.

Though the `redis-rb` client does not do this so we have to do this manually.

Reply:
Sorry, it's `script load` that returns the SHA.

Reply:
Right, I still don't like it. If we calculate the SHA differently from what Redis calculates (like a stripped linefeed), it will always be wrong and we will not error. Please use script load to be safe.

Reply:
Sure. I'll try to get back to this tomorrow.

Reply:
@mperham @engwan What do you think about a new module `Sidekiq::Scripting` for dealing with lua stuff just like what Sidekiq Pro does? Just an idea for refactoring.

Reply:
If Sidekiq Pro already has a pattern for loading a script and running it using EVALSHA, it would be great to extract that.

I was also wondering if EVALSHA is worth it here. We could also just call EVAL every time since Redis also does some optimizations already. The script is short anyway.

But if Pro does that too and we could extract it into a common module, then it's a no-brainer to just use it.

Reply:
I have updated the current PR to use `SCRIPT LOAD`. Let me know if you think we should extract this into a separate module as suggested above.